// haredo/dist/consumer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyMiddleware = exports.makeConsumer = void 0;
const haredo_message_1 = require("./haredo-message");
const bluebird_1 = require("bluebird");
const events_1 = require("./events");
const errors_1 = require("./errors");
const haredo_1 = require("./haredo");
const utils_1 = require("./utils");
const message_manager_1 = require("./message-manager");
exports.makeConsumer = async (cb, connectionManager, opts, log) => {
    let channel;
    let messageManager = message_manager_1.makeMessageManager(log);
    let consumerTag;
    const emitter = events_1.makeEmitter();
    const { noAck, exclusive, priority } = opts;
    const close = async () => {
        consumer.isClosing = true;
        if (consumerTag && channel) {
            await channel.cancel(consumerTag);
        }
        await messageManager.drain();
        emitter.emit('close');
        consumer.isClosed = true;
    };
    const setPrefetch = (prefetch) => {
        return channel.prefetch(prefetch, true);
    };
    const start = async () => {
        await opts.setup();
        channel = await connectionManager.getChannel();
        channel.once('close', async () => {
            log.info({ component: 'Consumer', msg: 'channel closed' });
            channel = null;
            if (opts.reestablish) {
                await bluebird_1.delay(5);
                try {
                    if (!consumer.isClosing) {
                        messageManager = message_manager_1.makeMessageManager(log);
                        await start();
                        emitter.emit('reestablished');
                    }
                }
                catch (error) {
                    // TODO: attempt again
                    log.error({ component: 'Consumer', msg: 'Failed to restart consumer', error });
                    emitter.emit('error', error);
                }
            }
            else if (!consumer.isClosing) {
                await close();
            }
        });
        await setPrefetch(opts.prefetch || 0);
        ({ consumerTag } = await channel.consume(opts.queue.getName(), async (message) => {
            var _a, _b, _c, _d, _e;
            if (message === null) {
                return;
            }
            await ((_a = opts.backoff) === null || _a === void 0 ? void 0 : _a.take());
            let messageInstance;
            const methods = {
                ack: () => {
                    var _a, _b;
                    /* istanbul ignore if */
                    if (!channel) {
                        throw new errors_1.ChannelBrokenError();
                    }
                    if (!noAck) {
                        channel.ack(message);
                    }
                    (_b = (_a = opts.backoff) === null || _a === void 0 ? void 0 : _a.ack) === null || _b === void 0 ? void 0 : _b.call(_a);
                },
                nack: (requeue = true) => {
                    var _a, _b;
                    /* istanbul ignore if */
                    if (!channel) {
                        throw new errors_1.ChannelBrokenError();
                    }
                    if (!noAck) {
                        channel.nack(message, false, requeue);
                    }
                    (_b = (_a = opts.backoff) === null || _a === void 0 ? void 0 : _a.nack) === null || _b === void 0 ? void 0 : _b.call(_a, requeue);
                },
                reply: async (reply) => {
                    if (!(message.properties.replyTo && message.properties.correlationId)) {
                        return;
                    }
                    await haredo_1.initialChain({ connectionManager })
                        .queue(message.properties.replyTo)
                        .skipSetup()
                        .confirm()
                        .publish(opts.json ? JSON.stringify(reply) : reply, {
                        correlationId: message.properties.correlationId
                    });
                }
            };
            try {
                if (consumer.isClosing) {
                    return;
                }
                messageInstance = haredo_message_1.makeHaredoMessage(message, opts.json, opts.queue.getName(), methods);
                messageManager.add(messageInstance);
                await exports.applyMiddleware(opts.middleware || [], cb, messageInstance, opts.autoAck, opts.autoReply, log);
                (_c = (_b = opts.backoff) === null || _b === void 0 ? void 0 : _b.pass) === null || _c === void 0 ? void 0 : _c.call(_b);
            }
            catch (error) {
                (_e = (_d = opts.backoff) === null || _d === void 0 ? void 0 : _d.fail) === null || _e === void 0 ? void 0 : _e.call(_d, error);
                if (!messageInstance) {
                    log.error({ component: 'Consumer', error, msg: 'failed initializing a message instance', rawMessage: message });
                    methods.nack(false);
                }
                else {
                    log.error({ component: 'Consumer', error, msg: 'error while handling message', message: messageInstance, rawMessage: message });
                    if (!noAck) {
                        messageInstance.nack(true);
                    }
                }
            }
        }, { noAck, priority, exclusive }));
    };
    const consumer = {
        close,
        emitter,
        isClosed: false,
        isClosing: false,
        prefetch: setPrefetch
    };
    await start();
    return consumer;
};
exports.applyMiddleware = async (middleware, cb, msg, autoAck, autoReply, log) => {
    if (!middleware.length) {
        const response = await cb(msg);
        if (typeof response !== 'undefined' && autoReply) {
            await msg.reply(response);
        }
        if (autoAck) {
            msg.ack();
        }
    }
    else {
        let nextWasCalled = false;
        await utils_1.head(middleware)(msg, () => {
            nextWasCalled = true;
            if (msg.isHandled()) {
                log.warning({ component: 'Consumer', msg: 'message was handled in the middleware but middleware called next() anyway', message: msg, rawMessage: msg.raw });
                return;
            }
            return exports.applyMiddleware(utils_1.tail(middleware), cb, msg, autoAck, autoReply, log);
        });
        if (!nextWasCalled && !msg.isHandled()) {
            await exports.applyMiddleware(utils_1.tail(middleware), cb, msg, autoAck, autoReply, log);
        }
    }
};
