// haredo/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FailedParsingJsonError = exports.HaredoClosingError = exports.ChannelBrokenError = exports.InvalidOptionsError = exports.HaredoError = void 0;
class HaredoError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'HaredoError';
    }
}
exports.HaredoError = HaredoError;
class InvalidOptionsError extends HaredoError {
    constructor(key) {
        super(`Error, connection object contains unknown key: ${key}`);
        this.key = key;
        this.name = 'InvalidOptionsError';
    }
}
exports.InvalidOptionsError = InvalidOptionsError;
class ChannelBrokenError extends HaredoError {
    /* istanbul ignore next */
    constructor() {
        super('Cannot ack/nack message, channel is already closed');
    }
}
exports.ChannelBrokenError = ChannelBrokenError;
class HaredoClosingError extends HaredoError {
    /* istanbul ignore next */
    constructor() {
        super('Haredo is closing, cannot create new connection');
    }
}
exports.HaredoClosingError = HaredoClosingError;
class FailedParsingJsonError extends HaredoError {
    constructor(data) {
        super('Failed to parse JSON for message');
        this.data = data;
    }
}
exports.FailedParsingJsonError = FailedParsingJsonError;
