// haredo/dist/exchange.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHaredoExchange = exports.makeExchangeConfig = exports.standardDelayedTypesArray = void 0;
const XDELAYEDTYPEKEY = 'x-delayed-type';
exports.standardDelayedTypesArray = [
    'direct',
    'fanout',
    'topic',
    'headers'
];
/**
 * Create a Exchange object to aid in setup (note: this doesn't assert it)
 * @param name name of the exchange
 * @param type Set the exchange type (direct/topic/headers/fanout/x-delayed-message)
 * @param opts Options that will be passed directly to amqplib
 * [amqplib#assertExchange](https://www.squaremobius.net/amqp.node/channel_api.html#channel_assertExchange)
 */
exports.makeExchangeConfig = (name, type, opts = {}) => {
    const cloneOpts = (top) => (Object.assign(Object.assign(Object.assign({}, opts), top), { arguments: Object.assign(Object.assign({}, opts.arguments), top.arguments), preferences: Object.assign(Object.assign({}, opts.preferences), top.preferences) }));
    return {
        metaType: 'exchange',
        getName: () => name,
        getType: () => type,
        getOpts: () => cloneOpts({}),
        durable: (durable = true) => exports.makeExchangeConfig(name, type, cloneOpts({ durable })),
        autoDelete: (autoDelete = true) => exports.makeExchangeConfig(name, type, cloneOpts({ autoDelete })),
        alternateExchange: (alternateExchange) => {
            if (typeof alternateExchange !== 'string') {
                alternateExchange = alternateExchange.getName();
            }
            return exports.makeExchangeConfig(name, type, Object.assign({}, opts, { alternateExchange }));
        },
        direct: () => exports.makeExchangeConfig(name, 'direct', opts),
        fanout: () => exports.makeExchangeConfig(name, 'fanout', opts),
        headers: () => exports.makeExchangeConfig(name, 'headers', opts),
        topic: () => exports.makeExchangeConfig(name, 'topic', opts),
        delayed: (xDelayedType) => exports.makeExchangeConfig(name, 'x-delayed-message', cloneOpts({ arguments: { 'x-delayed-type': xDelayedType } })),
        passive: (passive = true) => exports.makeExchangeConfig(name, type, cloneOpts({ preferences: { passive } }))
    };
};
/**
 * Returns true if passed in object is an exchange. Acts as a type guard for Exchange.
 * @param obj Object to check
 */
exports.isHaredoExchange = (obj) => {
    return (obj === null || obj === void 0 ? void 0 : obj.metaType) === 'exchange';
};
