// haredo/dist/haredo-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHaredoMessage = exports.makeHaredoMessage = void 0;
const events_1 = require("./events");
const utils_1 = require("./utils");
exports.makeHaredoMessage = (raw, parseJson, queue, methods) => {
    const state = {
        isHandled: false,
        isAcked: false,
        isNacked: false,
        isReplied: false,
        reply: undefined,
    };
    const dataString = raw.content.toString();
    const data = parseJson ? utils_1.parseJSON(dataString) : dataString;
    const emitter = events_1.makeEmitter();
    return {
        emitter,
        raw,
        dataString,
        data,
        queue,
        isHandled: () => state.isHandled,
        isAcked: () => state.isAcked,
        isNacked: () => state.isNacked,
        isReplied: () => state.isReplied,
        getReply: () => state.reply,
        ack: () => {
            if (state.isHandled) {
                return;
            }
            state.isAcked = true;
            state.isHandled = true;
            methods.ack();
            emitter.emit('handled');
        },
        nack: (requeue = true) => {
            if (state.isHandled) {
                return;
            }
            state.isNacked = true;
            state.isHandled = true;
            methods.nack(requeue);
            emitter.emit('handled');
        },
        reply: (message) => {
            state.isReplied = true;
            state.reply = message;
            return methods.reply(message);
        },
        getHeader: (header) => raw.properties.headers[header],
        headers: raw.properties.headers,
        appId: raw.properties.appId,
        consumerTag: raw.fields.consumerTag,
        contentEncoding: raw.properties.contentEncoding,
        contentType: raw.properties.contentType,
        correlationId: raw.properties.correlationId,
        deliveryMode: raw.properties.deliveryMode,
        deliveryTag: raw.fields.deliveryTag,
        exchange: raw.fields.exchange,
        expiration: raw.properties.expiration,
        messageId: raw.properties.messageId,
        priority: raw.properties.priority,
        redelivered: raw.fields.redelivered,
        replyTo: raw.properties.replyTo,
        routingKey: raw.fields.routingKey,
        timestamp: raw.properties.timestamp,
        type: raw.properties.type,
        userId: raw.properties.userId,
        metaType: 'message',
        deliveryCount: raw.properties.headers['x-delivery-count']
    };
};
/**
 * Returns true if passed in object is an message. Acts as a type guard for Message.
 * @param obj Object to check
 */
exports.isHaredoMessage = (obj) => {
    return (obj === null || obj === void 0 ? void 0 : obj.metaType) === 'message';
};
