// haredo/dist/prepared-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHaredoPreparedMessage = exports.mergeHeaders = exports.mergeOptions = exports.mergeMessageState = exports.preparedMessage = void 0;
exports.preparedMessage = (state = {}) => {
    return {
        metaType: 'preparedMessage',
        getState: () => state,
        appId: (appId) => exports.preparedMessage(exports.mergeMessageState(state, { options: { appId } })),
        blindCarbonCopy: (...BCC) => exports.preparedMessage(exports.mergeMessageState(state, { options: { BCC } })),
        carbonCopy: (...CC) => exports.preparedMessage(exports.mergeMessageState(state, { options: { CC } })),
        contentEncoding: (contentEncoding) => exports.preparedMessage(exports.mergeMessageState(state, { options: { contentEncoding } })),
        contentType: (contentType) => exports.preparedMessage(exports.mergeMessageState(state, { options: { contentType } })),
        correlationId: (correlationId) => exports.preparedMessage(exports.mergeMessageState(state, { options: { correlationId } })),
        delay: (delay) => exports.preparedMessage(state).setHeader('x-delay', delay),
        expiration: (expiration) => exports.preparedMessage(exports.mergeMessageState(state, { options: { expiration } })),
        json: (content) => {
            const chain = exports.preparedMessage(state).contentType('application/json');
            if (content !== undefined) {
                return chain.rawContent(JSON.stringify(content));
            }
            return chain;
        },
        mandatory: (mandatory = true) => exports.preparedMessage(exports.mergeMessageState(state, { options: { mandatory } })),
        messageId: (messageId) => exports.preparedMessage(exports.mergeMessageState(state, { options: { messageId } })),
        persistent: (persistent = true) => exports.preparedMessage(exports.mergeMessageState(state, { options: { persistent } })),
        priority: (priority) => exports.preparedMessage(exports.mergeMessageState(state, { options: { priority } })),
        rawContent: (content) => exports.preparedMessage(exports.mergeMessageState(state, { content })),
        replyTo: (replyTo) => exports.preparedMessage(exports.mergeMessageState(state, { options: { replyTo } })),
        routingKey: (routingKey) => exports.preparedMessage(exports.mergeMessageState(state, { routingKey })),
        timestamp: (timestamp) => exports.preparedMessage(exports.mergeMessageState(state, { options: { timestamp } })),
        type: (type) => exports.preparedMessage(exports.mergeMessageState(state, { options: { type } })),
        userId: (userId) => exports.preparedMessage(exports.mergeMessageState(state, { options: { userId } })),
        setHeader: (header, value) => exports.preparedMessage(exports.mergeMessageState(state, { options: { headers: { [header]: value } } })),
        setHeaders: (headers) => exports.preparedMessage(exports.mergeMessageState(state, { options: { headers } }))
    };
};
exports.mergeMessageState = (base, top) => ({
    content: top.content || base.content,
    routingKey: top.routingKey || base.routingKey,
    options: exports.mergeOptions(base.options, top.options)
});
exports.mergeOptions = (base = {}, top = {}) => (Object.assign(Object.assign(Object.assign({}, base), top), { headers: exports.mergeHeaders(base.headers, top.headers) }));
exports.mergeHeaders = (base = {}, top = {}) => (Object.assign(Object.assign({}, base), top));
/**
 * Returns true if the passed in object is a preparedMessage chain. Acts as a type quard for MessageChain
 * @param obj item to check
 */
exports.isHaredoPreparedMessage = (obj) => {
    return (obj === null || obj === void 0 ? void 0 : obj.metaType) === 'preparedMessage';
};
