// haredo/dist/publisher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapChannel = exports.wrappedChannelGetter = exports.makePublisher = void 0;
const utils_1 = require("./utils");
exports.makePublisher = (cm, log) => {
    const getChannel = exports.wrappedChannelGetter(() => cm.getChannel(), log, false);
    const getConfirmChannel = exports.wrappedChannelGetter(() => cm.getConfirmChannel(), log, true);
    return {
        publishToExchange: async (exchange, routingKey, data, opts, confirm) => {
            let channel;
            if (confirm) {
                channel = await getConfirmChannel();
            }
            else {
                channel = await getChannel();
            }
            return channel.publishToExchange(exchange, routingKey, data, opts);
        },
        sendToQueue: async (queue, data, opts, confirm) => {
            let channel;
            if (confirm) {
                channel = await getConfirmChannel();
            }
            else {
                channel = await getChannel();
            }
            return channel.sendToQueue(queue, data, opts);
        }
    };
};
exports.wrappedChannelGetter = (channelGetter, log, isConfirmChannel) => {
    let channelPromise;
    let wrappedChannel;
    return async () => {
        if (wrappedChannel) {
            return wrappedChannel;
        }
        if (channelPromise) {
            await channelPromise;
            return wrappedChannel;
        }
        channelPromise = channelGetter();
        log.info({ component: 'Publisher', msg: 'opening channel' });
        const channel = await channelPromise;
        log.info({ component: 'Publisher', msg: 'channel opened' });
        wrappedChannel = exports.wrapChannel(channel, isConfirmChannel, log);
        channelPromise = undefined;
        channel.on('close', () => {
            log.info({ component: 'Publisher', msg: 'channel closed' });
            wrappedChannel.stop();
            wrappedChannel = undefined;
        });
        return wrappedChannel;
    };
};
exports.wrapChannel = (channel, confirm, log, ticketMachine = utils_1.makeTicketMachine()) => {
    channel.on('drain', () => {
        log.debug({ component: 'Publisher', msg: `${confirm ? 'confirm-channel' : 'channel'} drained, resuming publishing` });
        ticketMachine.play();
    });
    return {
        publishToExchange: async (exchange, routingKey, data, opts) => {
            const release = await ticketMachine.take();
            if (isConfirmChannel(channel, confirm)) {
                return new Promise((resolve, reject) => {
                    const ready = channel.publish(exchange, routingKey, data, opts, (err) => {
                        release();
                        if (err) {
                            return reject(err);
                        }
                        resolve();
                    });
                    if (!ready) {
                        ticketMachine.pause();
                    }
                });
            }
            const ready = channel.publish(exchange, routingKey, data, opts);
            if (!ready) {
                /* istanbul ignore next */
                log.debug({ component: 'Publisher', msg: `${confirm ? 'confirm-channel' : 'channel'} returned false on publishing, pausing publishing until drain` });
                ticketMachine.pause();
            }
            release();
        },
        sendToQueue: async (queue, data, opts) => {
            const release = await ticketMachine.take();
            if (isConfirmChannel(channel, confirm)) {
                return new Promise((resolve, reject) => {
                    const ready = channel.sendToQueue(queue, data, opts, (err) => {
                        release();
                        if (err) {
                            return reject(err);
                        }
                        resolve();
                    });
                    if (!ready) {
                        ticketMachine.pause();
                    }
                });
            }
            const ready = channel.sendToQueue(queue, data, opts);
            if (!ready) {
                /* istanbul ignore next */
                log.debug({ component: 'Publisher', msg: `${confirm ? 'confirm-channel' : 'channel'} returned false on publishing, pausing publishing until drain` });
                ticketMachine.pause();
            }
            release();
        },
        stop: () => {
            ticketMachine.stop();
        }
    };
};
const isConfirmChannel = (channel, isIt) => isIt;
