// haredo/dist/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHaredoQueue = exports.padString = exports.makeQueueConfig = exports.DEFAULT_QUEUE_OPTIONS = void 0;
exports.DEFAULT_QUEUE_OPTIONS = Object.freeze({
    durable: true,
    exclusive: false
});
/**
 * Create a queue configuration object. Note: this does not assert the queue
 *
 * @param name name of the queue, falsey value (including empty string) will make the server generate a name for you
 * @param opts queue options, passed to assertQueue
 * [amqplib#assertQueue](https://www.squaremobius.net/amqp.node/channel_api.html#channel_assertQueue)
 */
exports.makeQueueConfig = (name, opts = {}) => {
    const cloneOpts = (top) => (Object.assign(Object.assign(Object.assign({}, opts), top), { arguments: Object.assign(Object.assign({}, opts === null || opts === void 0 ? void 0 : opts.arguments), top === null || top === void 0 ? void 0 : top.arguments) }));
    const argument = (argName, value) => exports.makeQueueConfig(name, cloneOpts({ arguments: { [argName]: value } }));
    return {
        metaType: 'queue',
        getName: () => name,
        getOpts: () => cloneOpts(opts),
        durable: (durable = true) => exports.makeQueueConfig(name, cloneOpts({ durable })),
        autoDelete: (autoDelete = true) => exports.makeQueueConfig(name, cloneOpts({ autoDelete })),
        exclusive: (exclusive = true) => exports.makeQueueConfig(name, cloneOpts({ exclusive })),
        messageTtl: (messageTtl) => exports.makeQueueConfig(name, cloneOpts({ messageTtl })),
        maxLength: (maxLength) => exports.makeQueueConfig(name, cloneOpts({ maxLength })),
        expires: (expires) => exports.makeQueueConfig(name, cloneOpts({ expires })),
        dead: (deadLetterExchange, deadLetterRoutingKey) => {
            if (typeof deadLetterExchange !== 'string') {
                deadLetterExchange = deadLetterExchange.getName();
            }
            return exports.makeQueueConfig(name, cloneOpts({ deadLetterExchange, deadLetterRoutingKey }));
        },
        name: (name) => exports.makeQueueConfig(name, cloneOpts({})),
        mutateName: (newName) => { name = newName; },
        maxPriority: priority => argument('x-max-priority', priority),
        type: type => argument('x-queue-type', type),
        passive: (passive = true) => exports.makeQueueConfig(name, cloneOpts({ preferences: { passive } })),
        deliveryLimit: limit => argument('x-delivery-limit', limit),
        argument
    };
};
exports.padString = (str) => ` ${str} `;
/**
 * Returns true if passed in object is an queue. Acts as a type guard for Queue.
 * @param obj Object to check
 */
exports.isHaredoQueue = (obj) => {
    return (obj === null || obj === void 0 ? void 0 : obj.metaType) === 'queue';
};
