// haredo/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.walkUntilEnd = exports.makeTicketMachine = exports.parseJSON = exports.merge = exports.omitUndefined = exports.omit = exports.delay = exports.tail = exports.head = exports.promiseMap = void 0;
const errors_1 = require("./errors");
exports.promiseMap = async (arr, cb) => {
    return Promise.all(arr.map(cb));
};
exports.head = (arr) => arr[0];
exports.tail = (arr) => arr.slice(1);
exports.delay = (ms) => {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
};
exports.omit = (item, ...keys) => {
    const omittedItem = {};
    for (const [key, value] of Object.entries(item)) {
        if (!keys.includes(key)) {
            omittedItem[key] = value;
        }
    }
    return omittedItem;
};
exports.omitUndefined = (item) => {
    const undefinedKeys = Object.keys(item).filter((key) => {
        return item[key] === undefined;
    });
    return exports.omit(item, ...undefinedKeys);
};
exports.merge = (base, top) => {
    return Object.assign({}, base, top);
};
exports.parseJSON = (data) => {
    try {
        return JSON.parse(data);
    }
    catch (_a) {
        throw new errors_1.FailedParsingJsonError(data);
    }
};
exports.makeTicketMachine = () => {
    let paused = false;
    let waiting = false;
    let stopped = false;
    const tickets = [];
    const check = () => {
        if (stopped) {
            throw new Error('Ticketmachine has been stopped, no new tickets will be forthcoming');
        }
        if (paused || waiting) {
            return;
        }
        const ticket = tickets.shift();
        if (!ticket) {
            return;
        }
        waiting = true;
        ticket.resolve(() => {
            waiting = false;
            check();
        });
    };
    const play = () => {
        paused = false;
        check();
    };
    const pause = () => {
        paused = true;
    };
    const stop = (e = new Error()) => {
        stopped = true;
        tickets.forEach(x => x.reject(e));
        tickets.length = 0;
    };
    const take = async () => {
        let resolve;
        let reject;
        const promise = new Promise((res, rej) => {
            resolve = res;
            reject = rej;
        });
        tickets.push({ resolve, reject });
        check();
        return promise;
    };
    return {
        take,
        check,
        pause,
        play,
        stop
    };
};
exports.walkUntilEnd = (arr) => {
    let currentIndex = 0;
    return () => {
        const item = arr[currentIndex];
        if (currentIndex + 1 < arr.length) {
            currentIndex += 1;
        }
        return item;
    };
};
